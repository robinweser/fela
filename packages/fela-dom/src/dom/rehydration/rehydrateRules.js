/* @flow */
import { RULE_TYPE, generateDeclarationReference } from 'fela-utils'
import { camelCaseProperty } from 'css-in-js-utils'

import generateCacheEntry from './generateCacheEntry'

// Escaping for RegExp taken from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Escaping
function escapeRegExp(string) {
  return string.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&')
}

export default function rehydrateRules(
  css: string,
  media: string = '',
  support?: string = '',
  cache?: Object = {},
  specificityPrefix?: string = ''
): Object {
  let decl
  const DECL_REGEX = new RegExp(
    `${escapeRegExp(
      specificityPrefix
    )}[.]([0-9a-z_-]+)([^{]+)?{([^:]+):([^}]+)}`,
    'gi'
  )

  // This excellent parsing implementation was originally taken from Styletron and modified to fit Fela
  // https://github.com/rtsao/styletron/blob/master/packages/styletron-client/src/index.js#L47
  /* eslint-disable no-unused-vars,no-cond-assign */
  while ((decl = DECL_REGEX.exec(css))) {
    // $FlowFixMe
    const [ruleSet, className, pseudo, property, value] = decl
    /* eslint-enable */

    const declarationReference = generateDeclarationReference(
      camelCaseProperty(property),
      value,
      pseudo,
      media,
      support
    )

    cache[declarationReference] = generateCacheEntry(
      RULE_TYPE,
      className,
      property,
      value,
      pseudo,
      media,
      support,
      specificityPrefix
    )
  }

  return cache
}
