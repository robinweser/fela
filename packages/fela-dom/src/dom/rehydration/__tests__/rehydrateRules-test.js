import rehydrateRules from '../rehydrateRules'

describe('Rehydrating rules', () => {
  it('should rehydrate the renderer cache', () => {
    expect(
      rehydrateRules(
        '.a{color:red}.b{color:blue}.c:hover:active{border-color:2px solid rgb(255, 255, 0)}'
      )
    ).toMatchSnapshot()
  })
  it('should rehydrate the renderer cache with specifity prefix', () => {
    expect(
      rehydrateRules(
        '.parentClass .a{color:red}.parentClass .b{color:blue}.parentClass .c:hover:active{border-color:2px solid rgb(255, 255, 0)}',
        '',
        '',
        {},
        '.parentClass '
      )
    ).toMatchSnapshot()
  })
})
