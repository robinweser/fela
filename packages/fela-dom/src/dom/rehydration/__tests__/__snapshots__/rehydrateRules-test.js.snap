// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rehydrating rules should rehydrate the renderer cache 1`] = `
Object {
  ":hover:activeborderColor2px solid rgb(255, 255, 0)": Object {
    "className": "c",
    "declaration": "border-color:2px solid rgb(255, 255, 0)",
    "media": "",
    "pseudo": ":hover:active",
    "selector": ".c:hover:active",
    "support": "",
    "type": "RULE",
  },
  "colorblue": Object {
    "className": "b",
    "declaration": "color:blue",
    "media": "",
    "pseudo": "",
    "selector": ".b",
    "support": "",
    "type": "RULE",
  },
  "colorred": Object {
    "className": "a",
    "declaration": "color:red",
    "media": "",
    "pseudo": "",
    "selector": ".a",
    "support": "",
    "type": "RULE",
  },
}
`;

exports[`Rehydrating rules should rehydrate the renderer cache with specifity prefix 1`] = `
Object {
  ":hover:activeborderColor2px solid rgb(255, 255, 0)": Object {
    "className": "c",
    "declaration": "border-color:2px solid rgb(255, 255, 0)",
    "media": "",
    "pseudo": ":hover:active",
    "selector": ".parentClass .c:hover:active",
    "support": "",
    "type": "RULE",
  },
  "colorblue": Object {
    "className": "b",
    "declaration": "color:blue",
    "media": "",
    "pseudo": "",
    "selector": ".parentClass .b",
    "support": "",
    "type": "RULE",
  },
  "colorred": Object {
    "className": "a",
    "declaration": "color:red",
    "media": "",
    "pseudo": "",
    "selector": ".parentClass .a",
    "support": "",
    "type": "RULE",
  },
}
`;
